framework/libra-framework/sources/ol_sources/vouch.spec.move

```diff
@@ -0,0 +1,93 @@
+spec ol_framework::vouch {
+    use diem_framework::coin;
+    use ol_framework::libra_coin::LibraCoin;
+    use std::signer;
+    use diem_framework::system_addresses;
+    use ol_framework::epoch_helper;
+
+    spec module {
+        pragma verify = true;
+        pragma aborts_if_is_strict = false;
+
+        // Global invariant: ensure vectors are always aligned
+        invariant forall addr: address where exists<ReceivedVouches>(addr):
+            len(global<ReceivedVouches>(addr).incoming_vouches) == 
+            len(global<ReceivedVouches>(addr).epoch_vouched);
+    }
+
+
+
+    spec true_friends(addr: address): vector<address> {
+        pragma aborts_if_is_partial = true;
+        aborts_if false;
+        ensures forall a: address where vector::contains(result, a):
+            exists i: u64 where i < len(global<ReceivedVouches>(addr).incoming_vouches):
+                global<ReceivedVouches>(addr).incoming_vouches[i] == a;
+    }
+
+    spec is_valid_voucher_for(voucher: address, recipient: address): bool {
+        aborts_if false;
+        ensures result == vector::contains(true_friends(recipient), voucher);

```

Comment:
I'm not 100% what is being checked here, it may be I forgot how it this check works. Can you add a comment?